front/src/applications/operationalStudies/hooks/useLazySimulateTrains.ts

```diff
@@ -0,0 +1,92 @@
+import { useRef, useEffect, useCallback, useState } from 'react';
+
+import type {
+  LightRollingStockWithLiveries,
+  SimulationSummaryResult,
+} from 'common/api/osrdEditoastApi';
+import type { TimetableItemWithDetails } from 'modules/trainschedule/components/Timetable/types';
+import type { TimetableItemId, TimetableItemWithTimetableId } from 'reducers/osrdconf/types';
+import { useAppDispatch } from 'store';
+
+import formatTimetableItemSummaries from '../helpers/formatTimetableItemSummaries';
+import TrainSimulationLazyLoader from '../helpers/TrainSimulationLazyLoader';
+
+type UseLazySimulateTrainsOptions = {
+  infraId: number;
+  electricalProfileSetId: number | undefined;
+  rollingStocks: LightRollingStockWithLiveries[] | null;
+  onProgress?: (results: Map<TimetableItemId, TimetableItemWithDetails>) => void;
+};
+
+export default function useLazySimulateTrains({
+  infraId,
+  electricalProfileSetId,
+  rollingStocks,
+  onProgress,
+}: UseLazySimulateTrainsOptions) {
+  const dispatch = useAppDispatch();
+  const loaderRef = useRef<TrainSimulationLazyLoader>();
+  const timetableItemsByIdRef = useRef<Map<TimetableItemId, TimetableItemWithTimetableId>>(
+    new Map()
+  );
+  const [simulatedTrainsById, setSimulatedTrainsById] = useState<
+    Map<TimetableItemId, TimetableItemWithDetails>
+  >(new Map());
+
+  useEffect(() => {
+    if (!rollingStocks) return undefined;
+
+    const loader = new TrainSimulationLazyLoader({
+      dispatch,
+      infraId,
+      electricalProfileSetId,
+      onProgress: (rawSummaries: Map<TimetableItemId, SimulationSummaryResult>) => {
+        const summaries = formatTimetableItemSummaries(
+          [...rawSummaries.keys()],
+          rawSummaries,
+          timetableItemsByIdRef.current,
+          rollingStocks
+        );
+        setSimulatedTrainsById((prev) => new Map([...prev.entries(), ...summaries.entries()]));
+        if (onProgress) onProgress(summaries);
+      },
+    });
+
+    loader.simulateTimetableItems([...timetableItemsByIdRef.current.keys()]);
+
+    loaderRef.current = loader;
+    return () => {
+      loader.cancel();
+      loaderRef.current = undefined;
```

Comment:
Why do you need to set the current ref to undefined here ? Is it for  `allTrainsSimulated` ?

Reply:
Mostly a matter of avoiding a ref left behind with a loader object that we shouldn't use anymore.